### DISCUSSION TITLE
How to send emails from Supabase

### DISCUSSION THREAD

**Discussion by divinemaredi** [supabase/supabase#424] [Questions] [answered]
I'd like to know what's the recommended method of how to enable my backend to send emails when certain events occur?  e.g. when a user signs up

**Comment by kiwicopple** [comment, marked as answer]
Hey @divinemaredi - at the moment we don't have anything for this. But we're working on it!

For now you might want to create a server that listens to your database changes and sends emails using a transactional mail service (like mailgun, sendgrid, or mandrill).

**Reply by lienista** [reply]
Do you have anything yet, or should I go ahead w integrating w AWS?

**Reply by codeofsumit** [reply]
webhooks would be great!

**Reply by meetbryce** [reply]
any update on this?

**Comment by divinemaredi** [comment]
Just to clarify, the email functionality that you're working on, will it be able to verify a user's email address? And send a password reset email? And if so, when will this feature be available? In the beta release? Or when Supabase is production-ready? 

Thanks again!

**Reply by kiwicopple** [reply]
I don't have timelines yet for this one just yet sorry as we're working on quite a few features right now.

Just a note though, we already provide functionality for email confirmation and signup :)

<img width="1430" alt="image" src="https://user-images.githubusercontent.com/10214025/102882943-24e38c00-448a-11eb-8438-506abe28f78f.png">

**Reply by divinemaredi** [reply]
Thanks for bringing that up, just played around with it and I understand that the Site URL (In Authentification >> Settings) is the link that users will be redirected to from a "Password Reset" email, and this link needs to lead to a form.

My question is, on the form the user will land on to set a new password, what API request should I use to update the user's password? I reckon that, if using JS, I would use auth.update(), but I couldn't find a cURL alternative to auth.update() in the documentation.

That's why I'd kindly like to know which cURL request should I use to update a user's info? Thanks again!

**Reply by kiwicopple** [reply]
> if using JS, I would use auth.update(),

Here are the docs for Password Reset: https://supabase.io/docs/client/reset-password-email

And if using cURL it would be `POST /recover` - found on GoTrue https://github.com/netlify/gotrue. The full URL would be something like `https://your-url.supabase.co/auth/v1/recover`

Hope that helps!

**Reply by divinemaredi** [reply]
Thanks for the GoTrue link! I'd like to use cURL and I saw that to update a user's password, I need to use a `PUT /user` request. But the documentation doesn't reveal the Headers required.
![Update password](https://user-images.githubusercontent.com/65621952/102999199-477cb000-4531-11eb-9bd8-c212d3aec5f0.PNG)

So when I try this request on Postman using the anon key, I keep getting this 401 error:
`{
    "code": 401,
    "msg": "Invalid token: token contains an invalid number of segments"
}`

So I'd kindly like to know exactly what headers I should use when making this request, thanks!

**Reply by kiwicopple** [reply]
Perhaps you need to send it as `form/encoded` data rather than `json` data?

**Reply by divinemaredi** [reply]
Okay thanks a lot!

**Comment by MarcVanc** [comment]
Hi, any new developments on this.

Thqnk you,
Marc

**Reply by ashleyconnor** [reply]
It should be possible when the functions feature lands but not in `PL/v8` as the v8 runtime is a "trusted" runtime so doesn't allow HTTP calls.

**Reply by touhidrahman** [reply]
Now that functions (alpha) is released, would have been nice if Supabase added a guide on how to do this.

**Comment by johndpope** [comment]
another use case could be - if a user submits an application form - is there anyway to hook into this bespoke event?

UPDATE: workaround using ses + aws 
https://gist.github.com/johndpope/6c9d64687e9af008bff367b0864e9793

**Comment by tanifort** [comment]
@kiwicopple 
What is the best workflow to achieve Password-Reset for a React Native/ Mobile applications? Being that, an email can't always be received on the device where the application is installed. It is possible to compute the so-called `access_token `internally without having the user click on the recovery Link?

**Comment by indymaat** [comment]
I would like to send invoices based on received payment webhooks on an edge function. What would be a good approach here? Is it even possible to send emails with an edge function? Or should I resort to a 3rd party via an HTTP call to do the mailing?

**Reply by sturmenta** [reply]
now supabase [can send emails using `resend.com`](https://supabase.com/docs/guides/functions/examples/send-emails)

> resend give 3000 free emails / month

1) upload script to edge function
2) call it and send custom emails

they made really easy ⚡️

```javascript
import { serve } from 'https://deno.land/std@0.168.0/http/server.ts'

const RESEND_API_KEY = Deno.env.get('RESEND_API_KEY')

const handler = async (_request: Request): Promise<Response> => {
  const res = await fetch('https://api.resend.com/emails', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${RESEND_API_KEY}`,
    },
    body: JSON.stringify({
      from: 'onboarding@resend.dev',
      to: 'delivered@resend.dev',
      subject: 'hello world',
      html: '<strong>it works!</strong>',
    }),
  })

  const data = await res.json()

  return new Response(JSON.stringify(data), {
    status: 200,
    headers: {
      'Content-Type': 'application/json',
    },
  })
}

serve(handler)
```

**Reply by indymaat** [reply]
Thanks, this is indeed very simple and intuitive. Setting up the domain is also relatively simple. Now you need a sensible way for creating and handling HTML email templates.
